be/src/storage/persistent_index.h

```diff
@@ -0,0 +1,191 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#pragma once
+
+#include <memory>
+
+#include "common/statusor.h"
+#include "storage/edit_version.h"
+#include "util/phmap/phmap.h"
+
+namespace starrocks {
+
+template <size_t KeySize>
+struct FixedKey {
+    uint8_t data[KeySize];
+};
+
+template <size_t KeySize>
+bool operator==(const FixedKey<KeySize>& lhs, const FixedKey<KeySize>& rhs) {
+    return memcmp(lhs.data, rhs.data, KeySize) == 0;
+}
+
+using IndexValue = uint64_t;
+static constexpr IndexValue NullIndexValue = -1;
+
+constexpr size_t PageSize = 4096;
+constexpr size_t PageHeaderSize = 64;
+constexpr size_t BucketPadding = 16;
+constexpr size_t BucketPerPage = 16;
+
+template <class T, class P>
+T npad(T v, P p) {
+    return (v + p - 1) / p;
+}
+
+template <class T, class P>
+T pad(T v, P p) {
+    return npad(v, p) * p;
+}
+
+struct IndexHash {
+    IndexHash(uint64_t hash) : hash(hash) {}
+    uint64_t hash;
+    uint64_t shard() const { return hash >> 48; }
+    uint64_t page() const { return (hash >> 16) & 0xffffffff; }
+    uint64_t bucket() const { return (hash >> 8) & (BucketPerPage - 1); }
+    uint64_t tag() const { return hash & 0xff; }
+};
+
+struct L1Checks {
+    std::vector<uint32_t> key_idxes;
+    std::vector<uint64_t> hashes;
+};
+
+class IndexL0 {
```

Comment:
Is it better to name `IndexL0` as `MutableIndex` and name `IndexL1` as `ImmutableIndex`

Reply:
mutable & immutable do not imply the level of the data structure, just part of it's traits

Reply:
The info of level can represented by the owner. Say if you have three level, L1 and L2 shares the same code.

Reply:
OK, changed